File: edit_distance/edit_distance.rb
require 'pry'

def edit_distance(s1, s2)
  s1 = s1.chars
  s2 = s2.chars
  memo = Array.new(s1.size + 1){ Array.new(s2.size + 1) }
  (0..(s1.size)).each do |i|
    (0..(s2.size)).each do |j|
      if i.zero?
        memo[i][j] = j
      elsif j.zero?
        memo[i][j] = i
      else
        if s1[i] != s2[j]
          memo[i][j] = [memo[i-1][j] + 1, memo[i][j-1] + 1, memo[i-1][j-1] + 1].min
        else
          memo[i][j] = memo[i-1][j-1]
        end
      end
    end
  end
  # binding.pry
  p memo.last.last
  # p memo
end

edit_distance("algorithm", "altruistic")
edit_distance("saturday", "sunday")

File: edit_distance/edit_distance.ex
defmodule EditDistance do

  def edit_distance(string1, string2) do
    s1 = String.codepoints string1
    s2 = String.codepoints string2

    IO.inspect EditDistance.edit(s1, s2)
  end

  def edit(s1, s2) when length(s1) == 0 do
    length(s2)
  end

  def edit(s1, s2) when length(s2) == 0 do
    length(s1)
  end

  def edit(s1, s2) do
    [head1|tail1] = s1
    [head2|tail2] = s2
    # Insertion insert "head1" to s2
    insert = EditDistance.edit(tail1, s2) + 1
    # Deletetion delete head2 from s2
    delete = EditDistance.edit(s1, tail2) + 1
    # Substitute head2 to head1
    edit = if head1 == head2 do
             # If head1 == head2 edit cost is free
             # (you don't need to edit, just skip to next recursion)
             EditDistance.edit(tail1, tail2)
           else
             EditDistance.edit(tail1, tail2) + 1
           end
    costs = [insert, delete, edit]
    Enum.min(costs)
  end
end

EditDistance.edit_distance("algorithm", "altruistic")
EditDistance.edit_distance("saturday", "sunday")
